packages/type-utils/src/TypeOrValueSpecifier.ts

```diff
@@ -0,0 +1,202 @@
+import path from 'path';
+import type * as ts from 'typescript';
+
+interface FileSpecifier {
+  from: 'file';
+  name: string | Array<string>;
+  source?: string;
+}
+
+interface LibSpecifier {
+  from: 'lib';
+  name: string | Array<string>;
+}
+
+interface PackageSpecifier {
+  from: 'package';
+  name: string | Array<string>;
+  source: string;
+}
+
+export type TypeOrValueSpecifier =
+  | string
+  | FileSpecifier
+  | LibSpecifier
+  | PackageSpecifier;
+
+export const typeOrValueSpecifierSchema = {
+  oneOf: [
+    {
+      type: 'string',
+    },
+    {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        from: {
+          type: 'string',
+          const: 'file',
+        },
+        name: {
+          oneOf: [
+            {
+              type: 'string',
+            },
+            {
+              type: 'array',
+              minItems: 1,
+              uniqueItems: true,
+              items: {
+                type: 'string',
+              },
+            },
+          ],
+        },
+        source: {
+          type: 'string',
+        },
+      },
+      required: ['from', 'name'],
+    },
+    {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        from: {
+          type: 'string',
+          const: 'lib',
+        },
+        name: {
+          oneOf: [
+            {
+              type: 'string',
+            },
+            {
+              type: 'array',
+              minItems: 1,
+              uniqueItems: true,
+              items: {
+                type: 'string',
+              },
+            },
+          ],
+        },
+      },
+      required: ['from', 'name'],
+    },
+    {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        from: {
+          type: 'string',
+          const: 'package',
+        },
+        name: {
+          oneOf: [
+            {
+              type: 'string',
+            },
+            {
+              type: 'array',
+              minItems: 1,
+              uniqueItems: true,
+              items: {
+                type: 'string',
+              },
+            },
+          ],
+        },
+        source: {
+          type: 'string',
+        },
+      },
+      required: ['from', 'name', 'source'],
+    },
+    {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        from: {
+          type: 'array',
+          minItems: 1,
+          uniqueItems: true,
+          items: {
+            type: 'string',
+            enum: ['file', 'lib', 'package'],
+          },
+        },
+        name: {
+          type: 'string',
+        },
+      },
+      required: ['from', 'name'],
+    },
+  ],
+};
```

Comment:
Heh, these show up kinda weird:

![Screenshot of the options from the deployed preview of /rules/prefer-readonly-parameter-types#options](https://user-images.githubusercontent.com/3335181/215883705-ac7036d7-cd41-49e7-91a4-053874d9cd5e.png)

I don't know that there's much we can do beyond filing an issue on `json-schema-to-typescript`.

Reply:
I presume you mean the `name: string | [string, ...string[]]` part? That is actually a pretty neat way of encoding the `minItems: 1` property in the array, I wouldn't have thought of that...

I can remove that, but I started with the strictest schema I could come up with and an empty array really doesn't make sense here IMHO

Reply:
Yeah specifically what I'm looking at is how these are defined in this file. Because `minItems` is there, that's the type that gets printed.

We pretty very rarely use `minItems` in our schema descriptions for the same reason that we rarely use `[T, ...T][]` in types: even if they're more accurate, they're inconvenient to read/write and a lot of users are more confused by them than anything else. I think you can remove them:

```diff
- minItems: 1,
```

Reply:
I'm fine with removing the `minItems` option... But do you actually as a human ever encounter the generated interfaces? I thought you wouldn't...

Reply:
How come we allow config like this:
```ts
{
  from: ['file', 'lib', 'package'],
  name: 'foo',
}
```

This seems super ambiguous, IMO, because it doesn't include all the information (like the `source` property).

What's the intended usecase for this style of config?
Is it just so people can tersely define something? If yes - I personally think we should lean towards verbose and clear config over terse and ambiguous.

Reply:
Hi,
that is the `ManySpecifiers` form from #6017 - looking at the discussions there, I see me and @JoshuaKGoldberg agreed to remove this form, however, I either forgot that or we agreed to that after I implemented this - I don't remember, either way, I'll remove it.

Reply:
Oh, I previously removed it from the implementation and just forgot to remove it from the schema - that should be fixed now.

Reply:
Sweet - looks much better now!
![image](https://user-images.githubusercontent.com/7462525/223625264-9993917c-8bc8-40d7-a37d-2bf3ae36b318.png)